### PR TITLE
Add auto include filter hooks for CSP nonce and sha-256

### PR DESCRIPTION
**Pull request by sid77**
This PR adds auto include filter hooks for managing CSP nonces and sha-256 hashes.
The exposed hooks signatures are:
 - CoreExtensions::IntercomRails::AutoInclude.csp_nonce_hook(controller)
 - CoreExtensions::IntercomRails::AutoInclude.csp_sha256_hook(controller, sha-256 entry)

So, for instance, implementing nonce support with https://github.com/twitter/secureheaders can be done as follow:
```
module CoreExtensions
  module IntercomRails
    module AutoInclude
      def self.csp_nonce_hook(controller)
        nonce = SecureHeaders.content_security_policy_script_nonce(controller.request)
        nonce
      end
    end
  end
end
```
or, for sha-256 support:
```
module CoreExtensions
  module IntercomRails
    module AutoInclude
      def self.csp_sha256_hook(controller, sha256)
        SecureHeaders.append_content_security_policy_directives(controller.request, {script_src: [sha256]})
      end
    end
  end
end
```

### CODE REVIEW

File: lib/intercom-rails/auto_include_filter.rb [hunk 1]
```diff
@@ -56,7 +65,16 @@ def intercom_script_tag_called_manually?
       end
 
       def intercom_script_tag
-        @script_tag ||= ScriptTag.new(:find_current_user_details => true, :find_current_company_details => true, :controller => controller, :show_everywhere => show_everywhere?)
+        # User defined method for applying a nonce to the inserted js tag when
+        # using CSP
+        if defined?(CoreExtensions::IntercomRails::AutoInclude.csp_nonce_hook) == 'method'
```
Comment (ciaranlee) [inline lib/intercom-rails/auto_include_filter.rb]: I feel like this bit could be DRYd up a bit, give me a few minutes...

Comment (ciaranlee) [inline lib/intercom-rails/auto_include_filter.rb]: I think this reads much clearer:
```ruby
options = {
  :find_current_user_details => true,
  :find_current_company_details => true,
  :controller => controller,
  :show_everywhere => show_everywhere?
}

if defined?(CoreExtensions::IntercomRails::AutoInclude.csp_nonce_hook) == 'method'
  options.merge!(:nonce => CoreExtensions::IntercomRails::AutoInclude.csp_nonce_hook(controller.request))
end

@script_tag = ScriptTag.new(options)
```

What do you think?

I don't think the memoization (`||=-`) is necessary. If it is then a spec will fail when we change

**Comment by sid77** [comment]
Addressed @ciaranlee comment, also switched to passing the whole `controller` in case more than just `request` is needed

File: lib/intercom-rails/auto_include_filter.rb (hunk 1 above)
Comment (stevenharman) [reply lib/intercom-rails/auto_include_filter.rb]: I know it's nearly two years later, but I just hit a case where _not_ memoizing the `@script_tag` is having an impact. Specially, it means the `ScriptTag` ends up calling `ScriptTag#find_current_user_details` twice (b/c we initialize the `script_tag` twice in the `AutoIncludeFilter::filter` method). That results in running all `Proc`s used to configure the current user, customer, and any details. Which means we're doing some work twice. And some of that work might be DB calls.

So... any reason _not_ to memoize here?

**Pull request merged**